[New Command] Move-DbaDbFile

ClaudioESSilva opened the pull request:
<!-- Below information IS REQUIRED with every PR -->
## Type of Change
<!-- What type of change does your code introduce -->
 - [ ] Bug fix (non-breaking change, fixes #<!--issue number--> )
 - [x] New feature (non-breaking change, adds functionality, fixes #1806 )
 - [ ] Breaking change (effects multiple commands or functionality, fixes #<!--issue number--> )
 - [x] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Adding code coverage to existing functionality
 - [x] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/sqlcollaborative/appveyor-lab ?
 - [ ] Nunit test is included
 - [ ] Documentation
 - [ ] Build system
 
<!-- Below this line you can erase anything that is not applicable -->
### Purpose
<!-- What is the purpose or goal of this PR? (doesn't have to be an essay) --> 
A new command that allows move database files between paths.
Useful when we need to reorganize drives/mount points space or move a database to different paths.

### IMPORTANT
All my tests until now were done locally. 
Currently, I don't have access to an environment with remote instances to test the command.
If you are reading this and you can test it outside of the server, please let me know the results.
As soon as I can, I will also do those tests.

### Approach
<!-- How does this change solve that purpose -->
We need to indicate which database and the logical name we want to move and to where.
1 - Database will be set offline.
2 - File(s) copied
3 - For each file copied the metadata will be updated to point to the new path
4 - Optional - The deletion of the source files just happens if the `-DeleteAfterMove` switch is specified.
5 - Database set online

The destination needs to be accessible by SQL Server. Using `Test-DbaPath` to validate that.

#### 3 Major call types
1 - Move files by type (Data / Log or both)
``` powershell
Move-DbaDbFile -SqlInstance sql2017 -Database dbatools -FileType Data -FileDestination "D:\DATA2"
``` 
![image](https://user-images.githubusercontent.com/19521315/99463138-04755e80-292d-11eb-8cc1-788524f392e4.png)

2 - Specify which file you want to move using its logical name and the destination.
``` powershell
$fileToMove=@{
    'dbatools'='D:\DATA3'
    'dbatools_log'='D:\LOG2'
}
Move-DbaDbFile -SqlInstance sql2019 -Database dbatools -FileToMove $fileToMove
```
While copying
![image](https://user-images.githubusercontent.com/19521315/99462945-a5174e80-292c-11eb-89bf-36a2148bf117.png)

When finished with success
![image](https://user-images.githubusercontent.com/19521315/99462985-b9f3e200-292c-11eb-8b9a-7c33eea0e538.png)

3 - If we need some help creating the hashtable with the mapping of logical name and current folders we can use
``` powershell
Move-DbaDbFile -SqlInstance sql2017 -Database dbatools -FileStructureOnly
```
![image](https://user-images.githubusercontent.com/19521315/99462751-4225b780-292c-11eb-9ba9-73ac6c9e5758.png)

### Commands to test
<!-- if these are the examples in the help just note it as such -->
`Move-DbaDbFile`

### Learning
<!-- Optional -->
~~To do the copy, `Start-BitsTransfer` is used. During my tests, I found that sometimes ACLs were not replicated.
This leads to the database can't be set online as the moved data file can't be read by SQL Server.
To solve this I have used `-ACLFlags DACL` .~~

> ~~Copy the DACL ("Access Control List") information with the file.~~


As the parameter `-ACLFlags` of `Start-BitsTransfer` is not available on every PS version, I have a fallback to `Get-Acl ... | Set-Acl ... ` approach

niphlod commented:
will test asap

wsmelton commented on functions/Move-DbaDbFile.ps1:
```diff
@@ -0,0 +1,366 @@
+        Moves database files from one location to another.
+
+    .DESCRIPTION
+        Moves database files from one location to another.
```
I would recommend defining location. 

SQL Server 2016 brought support for storing data files on Azure Blob storage service...are we going to support that in this command? If not I would recommend layout what location means.

Example:

> Move database files from one local drive or folder to another.

wsmelton replied:
Which to even move it to Azure Blob can only be done via backup/restore, but to prevent confusion or questions I'd say local drives/folders.

ClaudioESSilva replied:
I agree. I updated it.
Thank you, Shawn.

wsmelton commented on functions/Move-DbaDbFile.ps1:
```diff
@@ -0,0 +1,366 @@
+    .EXAMPLE
+        PS C:\> Move-DbaDbFile -SqlInstance sql2017 -Database dbatools -FileStructureOnly
+
+        Shows the current database file structure (without filenames).
```
I would show the output of this command to be a complete example...

ClaudioESSilva replied:
Done

wsmelton commented on functions/Move-DbaDbFile.ps1:
```diff
@@ -0,0 +1,366 @@
+    process {
+        if (Test-FunctionInterrupt) { return }
+
+        if (!$FileType -and !$FileToMove -and !$FileStructureOnly ) {
```
Please expand `!` to use `-not` for readability.

ClaudioESSilva replied:
Done

wsmelton commented on functions/Move-DbaDbFile.ps1:
```diff
@@ -0,0 +1,366 @@
+                    } else {
+                        $testPS = Test-PSRemoting -ComputerName $server.ComputerName -ErrorAction Stop
+                    }
+                    if (!($testPS)) {
```
Please use `-not`

ClaudioESSilva replied:
Done

ClaudioESSilva commented:
Found that `-ACLFlags` on `Start-BitsTransfer` is not available on every PS version. Changed to `Get-Acl -Path $physicalName | Set-Acl $destination`.

Found that `Test-PSRemoting` is not using `-UseSSL` by relying on our dbatools config. Added this.

ClaudioESSilva commented:
I was doing more testing and found two issues.

1 - When the database is already offline and we use the `-FileType Data` parameter, the selected filter is not being honoured.
2 - Old servers (win Server 2012) have `BitsTransfer` v1.0 which can't be used remotely. Error yield is: "The remote use of BITS is not supported". In this scenario, we need to fall back to `Copy-Item`. 
I will add the logic to the script block and this way one try should be enough.

I will also add an improvement. When running locally do not use `Invoke-Command2` (which will use PSRemoting)

wsmelton commented:
Oh please don't use `Copy-Item` 🥺 🙏 

Robocopy may be an option but I can't recall what version of Windows Server they started shipping it with. I know 2016+ includes but not sure about 2012/2012 R2.

wsmelton commented:
@niphlod are we good with this one?

If so I'll update the manifest so it gets added in next release.

wsmelton commented on functions/Move-DbaDbFile.ps1:
```diff
@@ -0,0 +1,438 @@
+    param (
+        [parameter(Mandatory, ValueFromPipeline)]
+        [DbaInstanceParameter]$SqlInstance,
+        [PSCredential]$SqlCredential,
```
Since remoting is involved to the server for supporting moving files over `Invoke-Command`, and use of `Start-BitsTransfer` this requires the need for `Credential` parameter as well.

Both commands accept a parameter of `-Credential` and the use of the `SqlCredential` should not be used for those calls.

wsmelton commented on functions/Move-DbaDbFile.ps1:
```diff
@@ -0,0 +1,438 @@
+                    $testPS = $false
+                    if ($SqlCredential) {
+                        # why does Test-PSRemoting require a Credential param ? this is ugly...
+                        $testPS = Test-PSRemoting -ComputerName $server.ComputerName -Credential $SqlCredential -ErrorAction Stop
```
This should pass `$Credential`

wsmelton replied:
The comment can be removed as well. The parameter is required because utilizing WSMan requires elevated rights by default on a Windows Server.

wsmelton commented on functions/Move-DbaDbFile.ps1:
```diff
@@ -0,0 +1,438 @@
+                        # let's try to resolve it to a more qualified name, without "cutting" knowledge about the domain (only $server.Name possibly holds the complete info)
+                        $Resolved = (Resolve-DbaNetworkName -ComputerName $server.Name).FullComputerName
+                        if ($SqlCredential) {
+                            $testPS = Test-PSRemoting -ComputerName $Resolved -Credential $SqlCredential -ErrorAction Stop
```
Should pass `$Credential`

wsmelton commented on functions/Move-DbaDbFile.ps1:
```diff
@@ -0,0 +1,438 @@
+
+                                        Get-Acl -Path $physicalName | Set-Acl $destination
+                                    }
+                                    Invoke-Command2 -ComputerName $ComputerName -Credential $SqlCredential -ScriptBlock $scriptBlock -ArgumentList $physicalName, $destination
```
This should pass `$Credential`

wsmelton commented on functions/Move-DbaDbFile.ps1:
```diff
@@ -0,0 +1,438 @@
+
+                                        try {
+                                            Write-Message -Level Verbose -Message "Try copying using Start-BitsTransfer with UNC paths."
+                                            Start-BitsTransfer -Source $physicalNameUNC -Destination $destinationUNC -ErrorAction Stop
```
Should likely allow for `$Credential` as well but could leave it alone until we start seeing the use of the command for this use case.

wsmelton commented on functions/Move-DbaDbFile.ps1:
```diff
@@ -0,0 +1,438 @@
+                                                $source = $args[0]
+                                                Remove-Item -Path $source -ErrorAction Stop
+                                            }
+                                            Invoke-Command2 -ComputerName $ComputerName -Credential $SqlCredential -ScriptBlock $scriptBlock -ArgumentList $physicalName
```
Should pass `$Credential`

wsmelton reviewed (changes requested):
- Suggestions on docs and messages
- Need to include the `Credential` parameter because we are triggering remote calls to the server.

niphlod commented:
we're good, let the users try it and let's see their usecases before overengineering it

sheldonhull commented:
Perfect timing. I have a gerry rigged hybrid version of using backup/restore and so on, and it's not elegant. It's pretty high risk option when the cleanup step gets involved if doing backup and restore vs move for example. 

Is this ready for usage as I'd love to give it a try to free up some space on disks if it's stable and just needing some real-world usage feedback. Much appreciate this getting some love. Wanted to contribute myself on it but no capacity at this time :-(

ClaudioESSilva commented:
@wsmelton thank you very much. I got underwater with other topics and only today was able to come back here.

ClaudioESSilva commented:
@sheldonhull I have already used once on work.
Like any other command please do your tests for your scenario. As @niphlod said we don't want to overengineer it, I would say it's ready for the most common scenarios. However, if you find anything you think can be improved let us know so we can discuss it.
